src/targets/gpu/jit/mlir.cpp

```diff
@@ -125,14 +125,56 @@ struct mlir_compiler : compiler<mlir_compiler>
                 [=](module& m, instruction_ref ins, const std::vector<operation>& ops) {
                     std::vector<instruction_ref> inputs = ins->inputs();
+
+                    // Tuple inputs not supported
+                    assert(std::all_of(inputs.begin(), inputs.end() - 1, [](auto i) {
+                        return i->get_shape().sub_shapes().empty();
+                    }));
+
+                    // Multiple output case (allocate ins will give a tuple)
+                    std::vector<instruction_ref> flat_inputs(inputs);
+                    bool multi_out = not flat_inputs.back()->get_shape().sub_shapes().empty();
+                    if(multi_out)
+                    {
+                        auto allocs = flat_inputs.back();
+                        flat_inputs.pop_back();
+                        auto sub_shape_idx = range(allocs->get_shape().sub_shapes().size());
+                        std::transform(sub_shape_idx.begin(),
+                                       sub_shape_idx.end(),
+                                       std::back_inserter(flat_inputs),
+                                       [&](int i) {
+                                           return m.insert_instruction(
+                                               ins,
+                                               migraphx::make_op("get_tuple_elem", {{"index", i}}),
+                                               allocs);
+                                       });
+                    }
+                    std::vector<instruction_ref> tuple_replacements;
+
                     for(const auto i : range(mco.prefill_indices.size()))
                     {
                         auto prefilled_ins = m.insert_instruction(
                             ins,
                             migraphx::make_op("hip::fill", {{"value", mco.prefill_values[i]}}),
-                            inputs[mco.prefill_indices[i]]);
-                        replace(inputs, inputs[mco.prefill_indices[i]], prefilled_ins);
+                            flat_inputs[mco.prefill_indices[i]]);
+                        if(not multi_out or mco.prefill_indices[i] < inputs.size() - 1)
+                        {
+                            replace(inputs, inputs[mco.prefill_indices[i]], prefilled_ins);
+                        }
+                        else
+                        {
+                            tuple_replacements.push_back(prefilled_ins);
+                        }
                     }
+
+                    if(multi_out and not tuple_replacements.empty())
+                    {
+                        // Add identity to make sure fill operations happen before kernel call
+                        tuple_replacements.insert(tuple_replacements.begin(), inputs.back());
+                        inputs.back() = m.insert_instruction(
+                            ins, migraphx::make_op("identity"), tuple_replacements);
+                    }
```

Comment:
I'm curious, why does adding an identity instruction ensure that the fill happens before the kernel?

Reply:
Because the fills are added as arguments to the identity operator. The identity operator only returns the first argument(which is the allocation) and it ignores the other operators. When the scheduling is ran, it still sees the inputs and will make sure to put the fill instructions before the identity.

Reply:
Why would the the fill possibly happen after the kernel without the identity?

Reply:
Without the identity you have 2 outputs from the allocate instruction on independent paths which can be scheduled in any order:
Without identity:
```
alloc -> get_tuple(0) -> fill
|
-> mlir_op -> ... <rest of the graph>
```

With:
```
alloc -> get_tuple(0) -> fill
|                         |
---------------------------------> Identity -> mlir_op
```

Reply:
I see now. Thanks, that makes a lot more sense!